File: serializable/serializable.py
from typing import List, Tuple

data_items = dict()
conflict_graph = dict()

class DataItem:
    last_writer: str | None = None
    last_readers: List[str] = []


def parse_operations(input_string: str):
    tokens = input_string.split(' ')
    return [(tokens[i], tokens[i+1], tokens[i+2]) for i in range(0, len(tokens), 3)]


def init_data_structures(operations: List[Tuple[str, str, str]]):
    for (_, transaction_id, data_item_id) in operations:
        if not transaction_id in conflict_graph:
            conflict_graph[transaction_id] = []
        if not data_item_id in data_items:
            data_items[data_item_id] = DataItem()

# TODO: Do not add reflexive cycles


def build_conflict_graph(operations: List[Tuple[str, str, str]]):
    for (op, transaction_id, data_item_id) in operations:
        
        if op == 'w':
            # All reads before this operation conflict with the current write
            for reader in data_items[data_item_id].last_readers:
                conflict_graph[reader] += transaction_id
            
            # The last write before this operation conflicts with the current write (transitive closure omitted)
            last_writer = data_items[data_item_id].last_writer
            if last_writer is not None:
                conflict_graph[last_writer] += transaction_id

            # Update last writer as the current transaction
            data_items[data_item_id].last_writer = transaction_id

            # Clear the list of last readers
            data_items[data_item_id].last_readers = []

        if op == 'r':
            # Add the current transaction to the last readers list
            data_items[data_item_id].last_readers += transaction_id
            last_writer = data_items[data_item_id].last_writer
            if last_writer is not None:
                print(f'last writer was {last_writer} which conflicted with {transaction_id}')
                conflict_graph[last_writer] += transaction_id
        


operations = parse_operations(input())
init_data_structures(operations)
build_conflict_graph(operations)

print(operations)

print('----')

for x in conflict_graph.items():
    print (x)

print("---")

for (k ,v) in data_items.items():
    print (k, v.last_writer, v.last_readers)

File: serializable/serializable.cpp
#include <iostream>
#include <map>
#include <optional>
#include <vector>

#include "item.hpp"
#include "operation.hpp"
#include "serializable.hpp"
#include "transaction.hpp"
#include "utils.hpp"

using item_states = std::map<item_id, item_state>;
using conflict_graph = std::map<t_id, std::vector<t_id>>;

namespace {
    bool has_cycle_(const conflict_graph& graph, std::map<t_id, bool>& visited, std::map<t_id, bool>& visiting, const t_id transaction) {
    //std::cout << "Visiting " << transaction << std::endl;
    if (!visited[transaction]) {
    visited[transaction] = true;
    visiting[transaction] = true;
    if (graph.count(transaction) > 0) {
        for (const auto& conflict : graph.at(transaction)) {
        //std::cout << "Visiting child " << conflict << std::endl;
        if (visiting[conflict] || !visited[conflict] && has_cycle_(graph, visited, visiting, conflict)) {
            return true;
        }
    }
    }
    
    }
    visiting[transaction] = false;
    return false;
}

bool has_cycle(const conflict_graph& graph) {

    const uint t_count = graph.size();
    std::map<t_id, bool> visited;
    std::map<t_id, bool> visiting;

    for (const auto& [transaction, conflicts] : graph) {
        if (!visited[transaction] && has_cycle_(graph, visited, visiting, transaction)) {
            return true;
        }
    }
    
    return false;
}

void add_conflict(conflict_graph& conflicts, const t_id src, const t_id dst) {
    if (src != dst) {
        conflicts[src].push_back(dst);
    }
}
}



bool is_serializable(operations_t history) {

  item_states items{};
  conflict_graph conflicts{};

  for (const operation_t &op : history) {
    std::visit(
        overload{[&conflicts, &items](const write_op &op) {
                   // All reads before this operation conflict with the current
                   // write
                   //std::cout << "Last readers of data item " << op.item << ":";
                   for (const t_id reader : items[op.item].last_readers) {
                    add_conflict(conflicts, reader, op.transaction);
                   }
                   //std::cout << std::endl;
                   // The last write before this operation conflicts with the
                   // current write
                   if (items[op.item].last_writer.has_value()) {
                    //std::cout << "Last writer " << items[op.item].last_writer.value() << " on data item " << op.item << " conflicts with current write by T_" << op.transaction << std::endl;
                     const t_id src = items[op.item].last_writer.value();
                     const t_id dst = op.transaction;
                     add_conflict(conflicts, src, dst);
                   }
                   // Update last writer as the current transaction
                   items[op.item].last_writer = op.transaction;

                   // Clear the list of last readers
                   items[op.item].last_readers.clear();
                 },
                 [&conflicts, &items](const read_op &op) {
                   // Add the current transaction to the last readers list
                   items[op.item].last_readers.push_back(op.transaction);
                   // The last write to the data item conflicts with the
                   // current write
                   if (items[op.item].last_writer.has_value()) {
                    //std::cout << "Last writer " << items[op.item].last_writer.value() << " on data item " << op.item << " conflicts with current read by T_" << op.transaction << std::endl;
                    const t_id src = items[op.item].last_writer.value();
                    const t_id dst = op.transaction;
                    add_conflict(conflicts, src, dst);
                   }
                 },
                 [&conflicts, &items](const commit_op &op) {
                   throw "Commit not implemented!";
                 },
                 [&conflicts, &items](const abort_op &op) {
                   throw "Abort not implemented!";
                 }},
        op);
  }
  return !has_cycle(conflicts);
}
